src/Discovery/Discovery.tsx

```diff
@@ -299,46 +299,60 @@ const Discovery: React.FunctionComponent<Props> = (props: Props) => {
-    // ---
-    const searchableFields = selectedSearchableTextFields.length > 0
-      ? selectedSearchableTextFields
-      : config.features.search.searchBar.searchableTextFields;
-    if (searchableFields) {
-      searchableFields.forEach((field) => {
-        const formattedFields = formatSearchIndex(field);
-        search.addIndex(formattedFields);
-      });
+    const cacheKey = JSON.stringify({
+      studies: props.studies,
+      fields: selectedSearchableTextFields,
+    });
+    // Check if the search object is already cached
+    if (searchCache[cacheKey]) {
+      setJsSearch(searchCache[cacheKey]);
+      props.onSearchChange(props.searchTerm); // Reinitialize search with the cached object
     } else {
-      config.studyColumns.forEach((column) => {
-        if (!column.contentType || column.contentType === 'string') {
-          const studyColumnFieldsArr = formatSearchIndex(column.field);
-          search.addIndex(studyColumnFieldsArr);
+      const search = new JsSearch.Search(config.minimalFieldMapping.uid);
+      search.indexStrategy = new JsSearch.AllSubstringsIndexStrategy();
+      // Choose which fields in the data to make searchable.
+      // If `searchableFields` are configured, enable search over only those fields.
+      // If `searchableAndSelectableTextFields` is configured and fields are selected,
+      //  enable search over only those fields.
+      // Otherwise, default behavior: enable search over all non-numeric fields
+      // in the table and the study description.
+      // ---
+      const searchableFields = selectedSearchableTextFields;
+      if (searchableFields.length > 0) {
+        searchableFields.forEach((field) => {
+          const formattedFields = formatSearchIndex(field);
+          search.addIndex(formattedFields);
+        });
+      } else {
+        config.studyColumns.forEach((column) => {
+          if (!column.contentType || column.contentType === 'string') {
+            const studyColumnFieldsArr = formatSearchIndex(column.field);
+            search.addIndex(studyColumnFieldsArr);
+          }
+        });
+        // Also enable search over preview field if present
+        if (config.studyPreviewField) {
+          const studyPreviewFieldArr = formatSearchIndex(
+            config.studyPreviewField.field,
+          );
+          search.addIndex(studyPreviewFieldArr);
         }
-      });
-      // Also enable search over preview field if present
-      if (config.studyPreviewField) {
-        const studyPreviewFieldArr = formatSearchIndex(config.studyPreviewField.field);
-        search.addIndex(studyPreviewFieldArr);
       }
+      search.addDocuments(props.studies);
+      // expose the search function
+      setJsSearch(search);
+      // Reinitialize search
+      props.onSearchChange(props.searchTerm);
+      // Cache the newly created search object
+      setSearchCache((prevSearchCache) => ({
+        ...prevSearchCache,
+        [cacheKey]: search,
```

Comment:
I think i get the idea of this change. But I'm not sure if we need to cache all the different searches, because I believe the way this code works now is it will cache all the combanations each time a different field selection has been applied.
I think the most noticeable slowness we are observing now happens when switching the selected search field back from a subset of seclections to the whole collection of them. Maybe we only need to cache the search that is for the whole collection of all searchable fields and reuse it. Could that be more efficient?

Reply:
While testing I noticed that the summary description field also caused slowness, so this approach helps with that too. Let me know if you want me to refactor this to only cache the whole collection.

Reply:
I'm still a big concerned about the size of this cached object since it would keeps growing. I don't know if you have tested this and to see if my concern is valid?

Reply:
That's a good point. I will investigate and circle back.